File: cloud-mobile-integration/firebase/ios-swift/FirestoreManager.swift
import Foundation
import FirebaseFirestore
import Combine

/// Firebase Firestore Manager for SwiftUI
/// Handles database operations with Firestore
@MainActor
class FirestoreManager: ObservableObject {
    static let shared = FirestoreManager()
    
    private let db = Firestore.firestore()
    private var listeners: [String: ListenerRegistration] = [:]
    
    @Published var isOnline: Bool = true
    
    private init() {
        setupNetworkListener()
    }
    
    /// Setup network listener
    private func setupNetworkListener() {
        db.addSnapshotsInSyncListener {
            print("All snapshots in sync")
        }
    }
    
    /// Enable offline persistence
    func enableOfflinePersistence() {
        let settings = FirestoreSettings()
        settings.isPersistenceEnabled = true
        settings.cacheSizeBytes = FirestoreCacheSizeUnlimited
        db.settings = settings
    }
    
    /// Enable network
    func enableNetwork() async throws {
        try await db.enableNetwork()
        isOnline = true
    }
    
    /// Disable network
    func disableNetwork() async throws {
        try await db.disableNetwork()
        isOnline = false
    }
    
    // MARK: - CRUD Operations
    
    /// Add document
    func addDocument<T: Encodable>(
        collection: String,
        data: T
    ) async throws -> String {
        let docRef = try db.collection(collection).addDocument(from: data)
        return docRef.documentID
    }
    
    /// Set document
    func setDocument<T: Encodable>(
        collection: String,
        documentId: String,
        data: T,
        merge: Bool = false
    ) async throws {
        if merge {
            try await db.collection(collection)
                .document(documentId)
                .setData(from: data, merge: true)
        } else {
            try await db.collection(collection)
                .document(documentId)
                .setData(from: data)
        }
    }
    
    /// Get document
    func getDocument<T: Decodable>(
        collection: String,
        documentId: String,
        as type: T.Type
    ) async throws -> T? {
        let snapshot = try await db.collection(collection)
            .document(documentId)
            .getDocument()
        
        return try snapshot.data(as: type)
    }
    
    /// Update document
    func updateDocument(
        collection: String,
        documentId: String,
        data: [String: Any]
    ) async throws {
        try await db.collection(collection)
            .document(documentId)
            .updateData(data)
    }
    
    /// Delete document
    func deleteDocument(
        collection: String,
        documentId: String
    ) async throws {
        try await db.collection(collection)
            .document(documentId)
            .delete()
    }
    
    /// Check if document exists
    func documentExists(
        collection: String,
        documentId: String
    ) async throws -> Bool {
        let snapshot = try await db.collection(collection)
            .document(documentId)
            .getDocument()
        
        return snapshot.exists
    }
    
    // MARK: - Query Operations
    
    /// Get all documents in collection
    func getAllDocuments<T: Decodable>(
        collection: String,
        as type: T.Type
    ) async throws -> [T] {
        let snapshot = try await db.collection(collection).getDocuments()
        return try snapshot.documents.compactMap { try $0.data(as: type) }
    }
    
    /// Query documents with conditions
    func queryDocuments<T: Decodable>(
        collection: String,
        as type: T.Type,
        queryBuilder: (Query) -> Query
    ) async throws -> [T] {
        let baseQuery = db.collection(collection)
        let query = queryBuilder(baseQuery)
        let snapshot = try await query.getDocuments()
        return try snapshot.documents.compactMap { try $0.data(as: type) }
    }
    
    /// Query with where clause
    func queryWhere<T: Decodable>(
        collection: String,
        field: String,
        isEqualTo value: Any,
        as type: T.Type
    ) async throws -> [T] {
        let snapshot = try await db.collection(collection)
            .whereField(field, isEqualTo: value)
            .getDocuments()
        
        return try snapshot.documents.compactMap { try $0.data(as: type) }
    }
    
    /// Query with pagination
    func queryWithPagination<T: Decodable>(
        collection: String,
        as type: T.Type,
        limit: Int,
        lastDocument: DocumentSnapshot? = nil,
        orderBy field: String = "createdAt",
        descending: Bool = true
    ) async throws -> (items: [T], lastSnapshot: DocumentSnapshot?) {
        var query = db.collection(collection)
            .order(by: field, descending: descending)
            .limit(to: limit)
        
        if let lastDocument = lastDocument {
            query = query.start(afterDocument: lastDocument)
        }
        
        let snapshot = try await query.getDocuments()
        let items = try snapshot.documents.compactMap { try $0.data(as: type) }
        let lastSnapshot = snapshot.documents.last
        
        return (items, lastSnapshot)
    }
    
    // MARK: - Real-time Listeners
    
    /// Listen to document changes
    func listenToDocument<T: Decodable>(
        collection: String,
        documentId: String,
        as type: T.Type
    ) -> AsyncThrowingStream<T?, Error> {
        AsyncThrowingStream { continuation in
            let listener = db.collection(collection)
                .document(documentId)
                .addSnapshotListener { snapshot, error in
                    if let error = error {
                        continuation.finish(throwing: error)
                        return
                    }
                    
                    guard let snapshot = snapshot else {
                        continuation.yield(nil)
                        return
                    }
                    
                    do {
                        let data = try snapshot.data(as: type)
                        continuation.yield(data)
                    } catch {
                        continuation.finish(throwing: error)
                    }
                }
            
            continuation.onTermination = { @Sendable _ in
                listener.remove()
            }
        }
    }
    
    /// Listen to collection changes
    func listenToCollection<T: Decodable>(
        collection: String,
        as type: T.Type,
        queryBuilder: ((Query) -> Query)? = nil
    ) -> AsyncThrowingStream<[T], Error> {
        AsyncThrowingStream { continuation in
            var query: Query = db.collection(collection)
            
            if let queryBuilder = queryBuilder {
                query = queryBuilder(query)
            }
            
            let listener = query.addSnapshotListener { snapshot, error in
                if let error = error {
                    continuation.finish(throwing: error)
                    return
                }
                
                guard let snapshot = snapshot else {
                    continuation.yield([])
                    return
                }
                
                do {
                    let items = try snapshot.documents.compactMap { 
                        try $0.data(as: type)
                    }
                    continuation.yield(items)
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            
            continuation.onTermination = { @Sendable _ in
                listener.remove()
            }
        }
    }
    
    /// Remove listener
    func removeListener(id: String) {
        listeners[id]?.remove()
        listeners.removeValue(forKey: id)
    }
    
    /// Remove all listeners
    func removeAllListeners() {
        listeners.values.forEach { $0.remove() }
        listeners.removeAll()
    }
    
    // MARK: - Batch Operations
    
    /// Perform batch write
    func batchWrite(
        operations: (WriteBatch) throws -> Void
    ) async throws {
        let batch = db.batch()
        try operations(batch)
        try await batch.commit()
    }
    
    /// Batch set documents
    func batchSetDocuments<T: Encodable>(
        collection: String,
        documents: [(id: String, data: T)]
    ) async throws {
        try await batchWrite { batch in
            for (id, data) in documents {
                let docRef = db.collection(collection).document(id)
                try batch.setData(from: data, forDocument: docRef)
            }
        }
    }
    
    /// Batch update documents
    func batchUpdateDocuments(
        collection: String,
        updates: [(id: String, data: [String: Any])]
    ) async throws {
        try await batchWrite { batch in
            for (id, data) in updates {
                let docRef = db.collection(collection).document(id)
                batch.updateData(data, forDocument: docRef)
            }
        }
    }
    
    /// Batch delete documents
    func batchDeleteDocuments(
        collection: String,
        documentIds: [String]
    ) async throws {
        try await batchWrite { batch in
            for id in documentIds {
                let docRef = db.collection(collection).document(id)
                batch.deleteDocument(docRef)
            }
        }
    }
    
    // MARK: - Transaction Operations
    
    /// Run transaction
    func runTransaction<T>(
        updateBlock: @escaping (Transaction) async throws -> T
    ) async throws -> T {
        return try await db.runTransaction { transaction, errorPointer in
            do {
                return try await updateBlock(transaction)
            } catch {
                errorPointer?.pointee = error as NSError
                return nil as! T
            }
        } as! T
    }
    
    // MARK: - Counter Operations
    
    /// Increment field
    func incrementField(
        collection: String,
        documentId: String,
        field: String,
        by value: Int = 1
    ) async throws {
        try await db.collection(collection)
            .document(documentId)
            .updateData([
                field: FieldValue.increment(Int64(value))
            ])
    }
    
    /// Decrement field
    func decrementField(
        collection: String,
        documentId: String,
        field: String,
        by value: Int = 1
    ) async throws {
        try await incrementField(
            collection: collection,
            documentId: documentId,
            field: field,
            by: -value
        )
    }
    
    // MARK: - Array Operations
    
    /// Add to array
    func arrayUnion(
        collection: String,
        documentId: String,
        field: String,
        elements: [Any]
    ) async throws {
        try await db.collection(collection)
            .document(documentId)
            .updateData([
                field: FieldValue.arrayUnion(elements)
            ])
    }
    
    /// Remove from array
    func arrayRemove(
        collection: String,
        documentId: String,
        field: String,
        elements: [Any]
    ) async throws {
        try await db.collection(collection)
            .document(documentId)
            .updateData([
                field: FieldValue.arrayRemove(elements)
            ])
    }
    
    // MARK: - Server Timestamp
    
    /// Set server timestamp
    func setServerTimestamp(
        collection: String,
        documentId: String,
        field: String
    ) async throws {
        try await db.collection(collection)
            .document(documentId)
            .updateData([
                field: FieldValue.serverTimestamp()
            ])
    }
    
    // MARK: - Helper Methods
    
    /// Get collection reference
    func collectionReference(_ path: String) -> CollectionReference {
        return db.collection(path)
    }
    
    /// Get document reference
    func documentReference(
        collection: String,
        documentId: String
    ) -> DocumentReference {
        return db.collection(collection).document(documentId)
    }
    
    /// Clear cache
    func clearCache() async throws {
        try await db.clearPersistence()
    }
}

// MARK: - Query Builder Extensions

extension FirestoreManager {
    /// Build complex query
    func buildQuery(
        collection: String,
        filters: [QueryFilter] = [],
        orderBy: [(field: String, descending: Bool)] = [],
        limit: Int? = nil
    ) -> Query {
        var query: Query = db.collection(collection)
        
        // Apply filters
        for filter in filters {
            query = filter.apply(to: query)
        }
        
        // Apply ordering
        for order in orderBy {
            query = query.order(by: order.field, descending: order.descending)
        }
        
        // Apply limit
        if let limit = limit {
            query = query.limit(to: limit)
        }
        
        return query
    }
}

/// Query filter
enum QueryFilter {
    case isEqualTo(field: String, value: Any)
    case isNotEqualTo(field: String, value: Any)
    case isLessThan(field: String, value: Any)
    case isLessThanOrEqualTo(field: String, value: Any)
    case isGreaterThan(field: String, value: Any)
    case isGreaterThanOrEqualTo(field: String, value: Any)
    case arrayContains(field: String, value: Any)
    case arrayContainsAny(field: String, values: [Any])
    case whereIn(field: String, values: [Any])
    case whereNotIn(field: String, values: [Any])
    
    func apply(to query: Query) -> Query {
        switch self {
        case .isEqualTo(let field, let value):
            return query.whereField(field, isEqualTo: value)
        case .isNotEqualTo(let field, let value):
            return query.whereField(field, isNotEqualTo: value)
        case .isLessThan(let field, let value):
            return query.whereField(field, isLessThan: value)
        case .isLessThanOrEqualTo(let field, let value):
            return query.whereField(field, isLessThanOrEqualTo: value)
        case .isGreaterThan(let field, let value):
            return query.whereField(field, isGreaterThan: value)
        case .isGreaterThanOrEqualTo(let field, let value):
            return query.whereField(field, isGreaterThanOrEqualTo: value)
        case .arrayContains(let field, let value):
            return query.whereField(field, arrayContains: value)
        case .arrayContainsAny(let field, let values):
            return query.whereField(field, arrayContainsAny: values)
        case .whereIn(let field, let values):
            return query.whereField(field, in: values)
        case .whereNotIn(let field, let values):
            return query.whereField(field, notIn: values)
        }
    }
}

File: cloud-mobile-integration/firebase/flutter/firestore_manager.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';

/// Firebase Firestore Manager for Flutter
/// Handles database operations with Firestore
class FirestoreManager {
  static final FirestoreManager _instance = FirestoreManager._internal();
  factory FirestoreManager() => _instance;
  FirestoreManager._internal();

  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final Map<String, StreamSubscription> _listeners = {};

  /// Enable offline persistence
  void enableOfflinePersistence() {
    _firestore.settings = const Settings(
      persistenceEnabled: true,
      cacheSizeBytes: Settings.CACHE_SIZE_UNLIMITED,
    );
  }

  /// Enable network
  Future<void> enableNetwork() async {
    await _firestore.enableNetwork();
  }

  /// Disable network
  Future<void> disableNetwork() async {
    await _firestore.disableNetwork();
  }

  // MARK: - CRUD Operations

  /// Add document
  Future<String> addDocument({
    required String collection,
    required Map<String, dynamic> data,
  }) async {
    final docRef = await _firestore.collection(collection).add(data);
    return docRef.id;
  }

  /// Set document
  Future<void> setDocument({
    required String collection,
    required String documentId,
    required Map<String, dynamic> data,
    bool merge = false,
  }) async {
    await _firestore
        .collection(collection)
        .doc(documentId)
        .set(data, SetOptions(merge: merge));
  }

  /// Get document
  Future<Map<String, dynamic>?> getDocument({
    required String collection,
    required String documentId,
  }) async {
    final snapshot =
        await _firestore.collection(collection).doc(documentId).get();

    return snapshot.data();
  }

  /// Update document
  Future<void> updateDocument({
    required String collection,
    required String documentId,
    required Map<String, dynamic> data,
  }) async {
    await _firestore.collection(collection).doc(documentId).update(data);
  }

  /// Delete document
  Future<void> deleteDocument({
    required String collection,
    required String documentId,
  }) async {
    await _firestore.collection(collection).doc(documentId).delete();
  }

  /// Check if document exists
  Future<bool> documentExists({
    required String collection,
    required String documentId,
  }) async {
    final snapshot =
        await _firestore.collection(collection).doc(documentId).get();

    return snapshot.exists;
  }

  // MARK: - Query Operations

  /// Get all documents in collection
  Future<List<Map<String, dynamic>>> getAllDocuments({
    required String collection,
  }) async {
    final snapshot = await _firestore.collection(collection).get();
    return snapshot.docs.map((doc) => doc.data()).toList();
  }

  /// Query documents with conditions
  Future<List<Map<String, dynamic>>> queryDocuments({
    required String collection,
    Query Function(Query query)? queryBuilder,
  }) async {
    Query query = _firestore.collection(collection);

    if (queryBuilder != null) {
      query = queryBuilder(query);
    }

    final snapshot = await query.get();
    return snapshot.docs
        .map((doc) => doc.data() as Map<String, dynamic>)
        .toList();
  }

  /// Query with where clause
  Future<List<Map<String, dynamic>>> queryWhere({
    required String collection,
    required String field,
    required dynamic value,
  }) async {
    final snapshot = await _firestore
        .collection(collection)
        .where(field, isEqualTo: value)
        .get();

    return snapshot.docs.map((doc) => doc.data()).toList();
  }

  /// Query with pagination
  Future<QueryResult> queryWithPagination({
    required String collection,
    int limit = 20,
    DocumentSnapshot? lastDocument,
    String orderByField = 'createdAt',
    bool descending = true,
  }) async {
    Query query = _firestore
        .collection(collection)
        .orderBy(orderByField, descending: descending)
        .limit(limit);

    if (lastDocument != null) {
      query = query.startAfterDocument(lastDocument);
    }

    final snapshot = await query.get();
    final items = snapshot.docs.map((doc) => doc.data() as Map<String, dynamic>).toList();
    final lastDoc = snapshot.docs.isNotEmpty ? snapshot.docs.last : null;

    return QueryResult(items: items, lastDocument: lastDoc);
  }

  // MARK: - Real-time Listeners

  /// Listen to document changes
  Stream<Map<String, dynamic>?> listenToDocument({
    required String collection,
    required String documentId,
  }) {
    return _firestore
        .collection(collection)
        .doc(documentId)
        .snapshots()
        .map((snapshot) => snapshot.data());
  }

  /// Listen to collection changes
  Stream<List<Map<String, dynamic>>> listenToCollection({
    required String collection,
    Query Function(Query query)? queryBuilder,
  }) {
    Query query = _firestore.collection(collection);

    if (queryBuilder != null) {
      query = queryBuilder(query);
    }

    return query.snapshots().map((snapshot) =>
        snapshot.docs.map((doc) => doc.data() as Map<String, dynamic>).toList());
  }

  /// Listen with listener ID for cancellation
  void listenWithId({
    required String listenerId,
    required String collection,
    required void Function(List<Map<String, dynamic>>) onData,
    void Function(Object error)? onError,
    Query Function(Query query)? queryBuilder,
  }) {
    Query query = _firestore.collection(collection);

    if (queryBuilder != null) {
      query = queryBuilder(query);
    }

    final subscription = query.snapshots().listen(
      (snapshot) {
        final items = snapshot.docs
            .map((doc) => doc.data() as Map<String, dynamic>)
            .toList();
        onData(items);
      },
      onError: onError,
    );

    _listeners[listenerId] = subscription;
  }

  /// Remove listener
  void removeListener(String listenerId) {
    _listeners[listenerId]?.cancel();
    _listeners.remove(listenerId);
  }

  /// Remove all listeners
  void removeAllListeners() {
    for (var subscription in _listeners.values) {
      subscription.cancel();
    }
    _listeners.clear();
  }

  // MARK: - Batch Operations

  /// Perform batch write
  Future<void> batchWrite({
    required void Function(WriteBatch batch) operations,
  }) async {
    final batch = _firestore.batch();
    operations(batch);
    await batch.commit();
  }

  /// Batch set documents
  Future<void> batchSetDocuments({
    required String collection,
    required List<DocumentData> documents,
  }) async {
    await batchWrite(operations: (batch) {
      for (var doc in documents) {
        final docRef = _firestore.collection(collection).doc(doc.id);
        batch.set(docRef, doc.data);
      }
    });
  }

  /// Batch update documents
  Future<void> batchUpdateDocuments({
    required String collection,
    required List<DocumentData> documents,
  }) async {
    await batchWrite(operations: (batch) {
      for (var doc in documents) {
        final docRef = _firestore.collection(collection).doc(doc.id);
        batch.update(docRef, doc.data);
      }
    });
  }

  /// Batch delete documents
  Future<void> batchDeleteDocuments({
    required String collection,
    required List<String> documentIds,
  }) async {
    await batchWrite(operations: (batch) {
      for (var id in documentIds) {
        final docRef = _firestore.collection(collection).doc(id);
        batch.delete(docRef);
      }
    });
  }

  // MARK: - Transaction Operations

  /// Run transaction
  Future<T> runTransaction<T>({
    required Future<T> Function(Transaction transaction) updateFunction,
  }) async {
    return await _firestore.runTransaction<T>(updateFunction);
  }

  /// Transfer balance example (transaction)
  Future<void> transferBalance({
    required String fromUserId,
    required String toUserId,
    required double amount,
  }) async {
    await runTransaction(updateFunction: (transaction) async {
      final fromRef = _firestore.collection('users').doc(fromUserId);
      final toRef = _firestore.collection('users').doc(toUserId);

      final fromSnapshot = await transaction.get(fromRef);
      final toSnapshot = await transaction.get(toRef);

      final fromBalance = fromSnapshot.data()?['balance'] as double? ?? 0.0;
      final toBalance = toSnapshot.data()?['balance'] as double? ?? 0.0;

      if (fromBalance < amount) {
        throw Exception('Insufficient balance');
      }

      transaction.update(fromRef, {'balance': fromBalance - amount});
      transaction.update(toRef, {'balance': toBalance + amount});
    });
  }

  // MARK: - Counter Operations

  /// Increment field
  Future<void> incrementField({
    required String collection,
    required String documentId,
    required String field,
    int value = 1,
  }) async {
    await _firestore.collection(collection).doc(documentId).update({
      field: FieldValue.increment(value),
    });
  }

  /// Decrement field
  Future<void> decrementField({
    required String collection,
    required String documentId,
    required String field,
    int value = 1,
  }) async {
    await incrementField(
      collection: collection,
      documentId: documentId,
      field: field,
      value: -value,
    );
  }

  // MARK: - Array Operations

  /// Add to array
  Future<void> arrayUnion({
    required String collection,
    required String documentId,
    required String field,
    required List<dynamic> elements,
  }) async {
    await _firestore.collection(collection).doc(documentId).update({
      field: FieldValue.arrayUnion(elements),
    });
  }

  /// Remove from array
  Future<void> arrayRemove({
    required String collection,
    required String documentId,
    required String field,
    required List<dynamic> elements,
  }) async {
    await _firestore.collection(collection).doc(documentId).update({
      field: FieldValue.arrayRemove(elements),
    });
  }

  // MARK: - Server Timestamp

  /// Set server timestamp
  Future<void> setServerTimestamp({
    required String collection,
    required String documentId,
    required String field,
  }) async {
    await _firestore.collection(collection).doc(documentId).update({
      field: FieldValue.serverTimestamp(),
    });
  }

  // MARK: - Helper Methods

  /// Get collection reference
  CollectionReference collectionReference(String path) {
    return _firestore.collection(path);
  }

  /// Get document reference
  DocumentReference documentReference({
    required String collection,
    required String documentId,
  }) {
    return _firestore.collection(collection).doc(documentId);
  }

  /// Clear cache
  Future<void> clearCache() async {
    await _firestore.clearPersistence();
  }

  /// Dispose
  void dispose() {
    removeAllListeners();
  }
}

/// Query result with pagination
class QueryResult {
  final List<Map<String, dynamic>> items;
  final DocumentSnapshot? lastDocument;

  QueryResult({
    required this.items,
    this.lastDocument,
  });
}

/// Document data helper
class DocumentData {
  final String id;
  final Map<String, dynamic> data;

  DocumentData({
    required this.id,
    required this.data,
  });
}

/// Query builder extensions
extension FirestoreQueryBuilder on Query {
  Query whereEquals(String field, dynamic value) {
    return where(field, isEqualTo: value);
  }

  Query whereNotEquals(String field, dynamic value) {
    return where(field, isNotEqualTo: value);
  }

  Query whereLessThan(String field, dynamic value) {
    return where(field, isLessThan: value);
  }

  Query whereLessThanOrEquals(String field, dynamic value) {
    return where(field, isLessThanOrEqualTo: value);
  }

  Query whereGreaterThan(String field, dynamic value) {
    return where(field, isGreaterThan: value);
  }

  Query whereGreaterThanOrEquals(String field, dynamic value) {
    return where(field, isGreaterThanOrEqualTo: value);
  }

  Query whereArrayContains(String field, dynamic value) {
    return where(field, arrayContains: value);
  }

  Query whereArrayContainsAny(String field, List<dynamic> values) {
    return where(field, arrayContainsAny: values);
  }

  Query whereIn(String field, List<dynamic> values) {
    return where(field, whereIn: values);
  }

  Query whereNotIn(String field, List<dynamic> values) {
    return where(field, whereNotIn: values);
  }

  Query orderByAscending(String field) {
    return orderBy(field, descending: false);
  }

  Query orderByDescending(String field) {
    return orderBy(field, descending: true);
  }
}
